File: tests/unit/protocol/bytes.test.ts
import { describe, expect, it } from 'vitest';
import {
  asciiToBytes,
  bytesToAscii,
  bytesToHex,
  bytesToUtf8,
  hexToBytes,
  tryHexToBytes,
  utf8ToBytes,
} from '../../../src/lib/protocol';

describe('protocol bytes', () => {
  it('encodes lowercase hex and decodes strict hex', () => {
    expect(bytesToHex(Uint8Array.from([0, 15, 255]))).toBe('000fff');
    expect(hexToBytes('000fff')).toEqual(Uint8Array.from([0, 15, 255]));
    expect(tryHexToBytes('AA')).toBeUndefined();
    expect(tryHexToBytes('abc')).toBeUndefined();
  });

  it('roundtrips UTF-8 and validates ASCII', () => {
    expect(bytesToUtf8(utf8ToBytes('hello \u2713'))).toBe('hello \u2713');
    expect(asciiToBytes('abc')).toEqual(Uint8Array.from([97, 98, 99]));
    expect(asciiToBytes('\u2713')).toBeUndefined();
    expect(bytesToAscii(Uint8Array.from([0x61, 0xff]))).toBeUndefined();
  });
});

File: crates/lkjstr-protocol/tests/bytes_test.rs
use lkjstr_protocol::{
    ascii_to_bytes, bytes_to_ascii, bytes_to_hex, bytes_to_utf8, hex_to_bytes, try_hex_to_bytes,
    utf8_to_bytes,
};

#[test]
fn encodes_lowercase_hex_and_decodes_strict_hex() -> Result<(), String> {
    assert_eq!(bytes_to_hex(&[0, 15, 255]), "000fff");
    assert_eq!(
        hex_to_bytes("000fff").map_err(|error| error.to_string())?,
        vec![0, 15, 255]
    );
    assert_eq!(try_hex_to_bytes("AA"), None);
    assert_eq!(try_hex_to_bytes("abc"), None);
    Ok(())
}

#[test]
fn roundtrips_utf8_and_validates_ascii() -> Result<(), String> {
    let bytes = utf8_to_bytes("hello \u{2713}");
    assert_eq!(
        bytes_to_utf8(&bytes).map_err(|error| error.to_string())?,
        "hello \u{2713}"
    );
    assert_eq!(ascii_to_bytes("abc"), Some(vec![97, 98, 99]));
    assert_eq!(ascii_to_bytes("\u{2713}"), None);
    assert_eq!(bytes_to_ascii(&[0x61, 0xff]), None);
    Ok(())
}
